results/resultaccumulator.go

```diff
@@ -0,0 +1,201 @@
+	}
+
+	for key, tag := range tagsMap {
+		// Convert CompOpString to ComparisonOperator
+		compOp, err := ComparisonOperatorString(strings.ToLower(tag.CompOpString))
+		if err != nil {
+			return errors.Wrapf(err, "invalid comparison operator for tag %s", key)
+		}
+		tag.CompOp = compOp
+		r.tagDB[key] = tag
+	}
+
+	return nil
+}
+
+func (r *ResultAccumulator) SubmitTag(_ context.Context, tagID string, value any) (bool, error) {
+	tag, ok := r.tagDB[tagID]
+	if !ok {
+		return false, errors.Errorf("tag not found: %s", tagID)
+	}
+
+	isPassing, err := tag.IsPassing(value)
+	if err != nil {
+		return false, errors.Wrapf(err, "failed to validate tag %s", tagID)
+	}
+
+	r.tagSubmissions[tagID] = TagSubmission{
+		Tag:       tag,
+		Value:     value,
+		IsPassing: isPassing,
+	}
+
+	if !isPassing {
+		r.allTagsPassing = false
+	}
+
+	return isPassing, nil
+}
+
+func (r *ResultAccumulator) SubmitError(_ context.Context, err error) error {
+	r.errorSubmissions = append(r.errorSubmissions, err)
+	r.allTagsPassing = false
+	return nil
+}
+
+func (r *ResultAccumulator) CompleteTest(_ context.Context, testID uuid.UUID, sequenceName string) (bool, error) {
+	dateTime := time.Now().Format("2006-01-02_15-04-05")
+
+	overallPassFail := r.allTagsPassing && len(r.errorSubmissions) == 0
+
+	if err := r.updateHistoricTests(testID.String(), sequenceName, dateTime, overallPassFail); err != nil {
+		return false, errors.Wrap(err, "failed to update historic tests")
+	}
+
+	for _, generator := range r.generators {
+		err := generator.Generate(
+			testID,
+			sequenceName,
+			r.tagSubmissions,
+			r.errorSubmissions,
+			overallPassFail,
+			r.reportsDir,
+		)
+		if err != nil {
+			return false, errors.Wrap(err, "failed to generate report")
+		}
+	}
+
+	// Reset cached submissions
+	r.tagSubmissions = make(map[string]TagSubmission)
+	r.errorSubmissions = []error{}
+	r.allTagsPassing = true
+
+	return overallPassFail, nil
+}
+
+func (r *ResultAccumulator) updateHistoricTests(testID, sequenceName, dt string, testPassed bool) error {
+	ymd, hms, _ := strings.Cut(dt, "_")
+
+	newTest := map[string]any{
+		"testId":       testID,
+		"sequenceName": sequenceName,
+		"testPassed":   testPassed,
+		"date":         ymd,
+		"time":         hms,
+	}
+
+	existingTests, err := r.loadExistingTests()
+	if err != nil {
+		return err
+	}
+
+	updatedTests := append([]map[string]any{newTest}, existingTests...)
+
+	return r.saveUpdatedTests(updatedTests)
+}
+
+func (r *ResultAccumulator) loadExistingTests() ([]map[string]any, error) {
+	data, err := os.ReadFile(r.historicTestsFP)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return []map[string]any{}, nil
+		}
+		return nil, errors.Wrap(err, "failed to read historic tests file")
+	}
+
+	var existingTests []map[string]any
+	err = json.Unmarshal(data, &existingTests)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to unmarshal historic tests")
+	}
+
+	return existingTests, nil
+}
```

Comment:
Does historicsTestsFP just point to a JSON of all previous local tests?

Reply:
yes, its how the old github pages system used to woek